staging/src/k8s.io/kubectl/pkg/cmd/apply/patcher.go

```diff
@@ -116,15 +116,23 @@ func (p *Patcher) patchSimple(obj runtime.Object, modified []byte, namespace, na
 		return nil, nil, errors.Wrapf(err, "retrieving original configuration from:\n%v\nfor:", obj)
 	}
 
+	var patchType types.PatchType
 	var patch []byte
+	if p.OpenapiSchema != nil {
+		// if openapischema is used, we'll try to get required patch type for this GVK from Open API.
+		// if it fails or could not find any patch type, fall back to baked-in patch type determination.
+		patchType, err = p.getPatchTypeFromOpenAPI(p.Mapping.GroupVersionKind)
+	}
 
-	patchType := types.StrategicMergePatchType
-	_, err = scheme.Scheme.New(p.Mapping.GroupVersionKind)
-	if err != nil {
-		if !runtime.IsNotRegisteredError(err) {
-			return nil, nil, errors.Wrapf(err, "getting instance of versioned object for %v:", p.Mapping.GroupVersionKind)
+	if p.OpenapiSchema == nil || err != nil {
+		patchType = types.StrategicMergePatchType
+		_, err = scheme.Scheme.New(p.Mapping.GroupVersionKind)
+		if err != nil {
+			if !runtime.IsNotRegisteredError(err) {
+				return nil, nil, errors.Wrapf(err, "getting instance of versioned object for %v:", p.Mapping.GroupVersionKind)
+			}
+			patchType = types.MergePatchType
```

Comment:
I'm still not happy that the determination of the patch type is not isolated in a specific place.

If you look ten lines below, in `case types.StrategicMergePatchType:`, we're doing almost the same thing we do in that if condition, i.e. `if openapi works, use SMP otherwise use json merge`. Why do we have to do it twice in two different locations?

Reply:
@apelisse  Actually you have already described the correct behavior pretty well in here https://github.com/kubernetes/kubernetes/pull/110454#issuecomment-1155451009 and now I re-read it carefully and did some changes. I hope, we are more aligned now for latest version of this patch.